### ISSUE TITLE
Make torch.linalg.{eigvals,eigvalsh,svdvals} differentiable

### ISSUE THREAD

**Issue by Lezcano** [pytorch/pytorch#56788]
## 🚀 Feature

The functions `eig`, `eigh`, `svd` and their derivatives present several problems.
- `eig` is not always well-defined (only when the matrix is diagonalisable)
- The eigen / singular vectors are not well-defined (if `v` is an eigen / singular vector, `-v`  is also an eigen / singular vector). Even more, when they have repeated eigen / singular values, the derivatives are not well defined (see https://github.com/pytorch/pytorch/issues/47599#issuecomment-812539850)

On the other hand, the eigenvalues / singular values of a matrix are always well-defined, and so are their derivatives when the eigenvalues / singular vectors are not used in the loss function.

As such, I see these functions as safer versions of the full factorisations, and should be preferred when possible. I believe we should support differentiability for these.

We could do so via a flag or, even better, detecting whether we are in a differentiable setting (if this is possible, I am not completely sure).

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @jianyuh @mruberry @heitorschueroff @walterddr @IvanYashchuk @xwang233

**Comment by mruberry** [comment]
Hey @Lezcano, thanks for filing this issue! Let's discuss it more at our linalg autograd sync, too.

Would one idea be to make eigvals() differentiable? We can detect, for example, if the input to the operation requires gradient. We already have some functions with different code paths depending on whether their inputs requires grad or not. 

And then the analogous question would be for eighvalsh() and svdvals(), right?

**Comment by Lezcano** [comment]
Yeah, that sounds like a good solution. 

And yes, let's discuss it in the autograd sync.

**Comment by Lezcano** [comment]
Fixed in https://github.com/pytorch/pytorch/pull/57188 https://github.com/pytorch/pytorch/pull/57276 https://github.com/pytorch/pytorch/pull/57189
